modules/querier/querier.go

```diff
@@ -167,40 +166,54 @@ func (q *Querier) FindTraceByID(ctx context.Context, req *tempopb.TraceByIDReque
 		for _, r := range responses {
 			trace := r.response.Trace
 			if trace != nil {
-				completeTrace, _, _, spanCount = tempo_util.CombineTraceProtos(completeTrace, trace)
-				if spanCount > 0 {
-					spanCountTotal = spanCount
-				}
+				completeTrace, _, _, spanCount = model.CombineTraceProtos(completeTrace, trace)
+				spanCountTotal += spanCount
+				traceCountTotal++
 			}
 		}
 		span.LogFields(ot_log.String("msg", "done searching ingesters"),
 			ot_log.Bool("found", completeTrace != nil),
-			ot_log.Int("spanCountTotal", spanCountTotal))
+			ot_log.Int("combinedSpans", spanCountTotal),
+			ot_log.Int("combinedTraces", traceCountTotal))
 	}
 
 	if req.QueryMode == QueryModeBlocks || req.QueryMode == QueryModeAll {
 		span.LogFields(ot_log.String("msg", "searching store"))
-		partialTraces, err := q.store.Find(opentracing.ContextWithSpan(ctx, span), userID, req.TraceID, req.BlockStart, req.BlockEnd)
+		partialTraces, dataEncodings, err := q.store.Find(opentracing.ContextWithSpan(ctx, span), userID, req.TraceID, req.BlockStart, req.BlockEnd)
 		if err != nil {
 			return nil, errors.Wrap(err, "error querying store in Querier.FindTraceByID")
 		}
 
 		span.LogFields(ot_log.String("msg", "done searching store"))
 
-		// combine partialTraces
-		for _, partialTrace := range partialTraces {
-			storeTrace := &tempopb.Trace{}
-			err = proto.Unmarshal(partialTrace, storeTrace)
-			if err != nil {
-				return nil, err
+		if len(partialTraces) != 0 {
+			traceCountTotal = 0
+			spanCountTotal = 0
+			// combine partialTraces
+			var allBytes []byte
+			baseEncoding := dataEncodings[0] // just arbitrarily choose an encoding. generally they will all be the same
+			for i, partialTrace := range partialTraces {
+				dataEncoding := dataEncodings[i]
+				allBytes, _, err = model.CombineTraceBytes(allBytes, partialTrace, baseEncoding, dataEncoding)
```

Comment:
IIUC, the encoding of `allBytes` is unpredictable here. If the two traces were equal in the first call to combine, it will remain in `baseEncoding`. Else it might be proto marshalled bytes

Reply:
Yes, it will be, but in the end we call:

`storeTrace, err := model.Unmarshal(allBytes, baseEncoding)` which will marshal whatever the encoding is into a *tracepb.Trace. So it will work.